File: javascript/src/Tags/ListTag.ts
import { Parser } from '..';
import { HtmlNode } from '../Nodes/HtmlNode';
import { INode } from '../Nodes/INode';
import { NodeCollection } from '../Nodes/NodeCollection';
import { ParseData } from '../ParseData';
import { State } from '../State';
import { Tag } from './Tag';

export class ListTag extends Tag {
    constructor() {
        super();
        this.Token = 'list';
        this.Element = 'ul';
        this.IsBlock = true;
    }

    public override Validate(options: Record<string, string>, text: string): boolean {
        const items = text.split('[*]')
            .map(x => x.trim())
            .filter(x => x?.length > 0);
        return super.Validate(options, text) && items.length > 0;
    }

    public override FormatResult(parser: Parser, data: ParseData, state: State, scope: string, options: Record<string, string>, text: string): INode {
        let before = '<' + this.Element;
        if (this.ElementClass != null) before += ' class="' + this.ElementClass + '"';
        before += '>\n';

        const content = new NodeCollection();
        const items = text.split('[*]')
            .map(x => x.trim())
            .filter(x => x?.length > 0);
        for (const item of items) {
            const node = new HtmlNode('<li>', parser.ParseTags(data, item, scope, this.TagContext()), '</li>\n');
            node.PlainBefore = '* ';
            node.PlainAfter = '\n';
            content.Nodes.push(node);
        }
        
        const after = '</' + this.Element + '>';
        const ret = new HtmlNode(before, content, after);
        ret.IsBlockNode = true;
        return ret;
    }
}

File: csharp/LogicAndTrick.WikiCodeParser/Tags/ListTag.cs
using System;
using System.Collections.Generic;
using System.Linq;
using LogicAndTrick.WikiCodeParser.Nodes;

namespace LogicAndTrick.WikiCodeParser.Tags
{
    public class ListTag : Tag
    {
        public ListTag()
        {
            Token = "list";
            Element = "ul";
            IsBlock = true;
        }

        public override bool Validate(Dictionary<string, string> options, string text)
        {
            var items = text.Split(new[] { "[*]" }, StringSplitOptions.None)
                .Select(x => x.Trim())
                .Where(x => x.Length > 0);
            return base.Validate(options, text) && items.Any();
        }

        public override INode FormatResult(Parser parser, ParseData data, State state, string scope, Dictionary<string, string> options, string text)
        {
            var before = "<" + Element;
            if (ElementClass != null) before += " class=\"" + ElementClass + "\"";
            before += ">\n";

            var content = new NodeCollection();
            var items = text.Split(new[] { "[*]" }, StringSplitOptions.None)
                .Select(x => x.Trim())
                .Where(x => x.Length > 0);
            foreach (var item in items)
            {
                content.Nodes.Add(new HtmlNode("<li>", parser.ParseTags(data, item, scope, TagContext), "</li>\n") { PlainBefore = "* ", PlainAfter = "\n" });
            }

            var after = "</" + Element + ">";
            return new HtmlNode(before, content, after) { IsBlockNode = true };
        }
    }
}
